subscription_manager/tests/test_controller.py

```diff
@@ -0,0 +1,82 @@
+import pytest
+from subscription_manager.exceptions import InvalidSubsFieldException, MissingFieldsException
+import subscription_manager.utils as utils
+from dateutil.relativedelta import relativedelta
+from subscription_manager.controller import Controller
+from subscription_manager.subscription import Subscription
+from subscription_manager.dbhelper import DBHelper
+from bson.objectid import ObjectId
+from unittest.mock import MagicMock
+from datetime import datetime
+
+
+@pytest.fixture
+def subscription_dict() -> dict:
+    """Returns generated random subscription with correct fields"""
+    return utils.subscription_generator()
+
+
+@pytest.fixture
+def mock_dbhelper() -> DBHelper:
+    """Returns mock for DBHelper to avoid writing to production database"""
+    mock = DBHelper
+    mock.add_subscription = MagicMock(return_value=ObjectId("5ef2081b329cba6d5b03b6ff"))
+    return mock
+
+
+@pytest.fixture
+def controller(mock_dbhelper) -> Controller:
+    """Returns Controller class instance"""
+    database = mock_dbhelper
+    return Controller(database)
+
+
+@pytest.mark.parametrize("subscription_dict", [utils.subscription_generator() for _ in range(5)])
+def test_add_subscription(subscription_dict, controller):
+    """Check successful subscription addition with correct parameters"""
+    result = controller.add_subscription(subscription_dict)
+    assert type(result[0]) == ObjectId
+    assert type(result[1] == Subscription)
+
+
+@pytest.mark.parametrize("paramNone", ['owner', 'name', 'frequency', 'start_date', 'price', 'currency', 'comment'])
+def test_add_subscription_with_none_fields(controller, subscription_dict, paramNone):
+    """Check subscribe addition with one of the field equals None"""
+    sub_dict = subscription_dict
+    sub_dict[paramNone] = None
+    with pytest.raises(InvalidSubsFieldException):
+        assert controller.add_subscription(sub_dict)
+
+
+def test_add_subscription_empty(controller):
+    """Check that Controller will not add empty subscription"""
+    with pytest.raises(MissingFieldsException):
+        assert controller.add_subscription({})
+
+
+@pytest.mark.parametrize("field, value, expected_error",
+                         [('owner', 123, InvalidSubsFieldException),
+                          ('name', 4.56, InvalidSubsFieldException),
+                          ('frequency', 31, InvalidSubsFieldException),
+                          ('start_date', 1012020, InvalidSubsFieldException),
+                          ('price', 'two', InvalidSubsFieldException),
+                          ('currency', 1.01, InvalidSubsFieldException),
+                          ('comment', int, InvalidSubsFieldException)])
```

Comment:
do we really need to test value == "int"

Reply:
done